File: algorithms/algorithms-cplusplus/leetcode/Question_0368_Largest_Divisible_Subset.cc
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> largestDivisibleSubset(vector<int>& nums) {
    sort(nums.begin(), nums.end());
    vector<int> dp(nums.size(), 0), parent(nums.size(), 0), res;
    int mx = 0, mx_idx = 0;
    for (int i = 0; i < nums.size(); i++) {
      for (int j = i; j >= 0; j--) {
        if (nums[i] % nums[j] == 0 && dp[i] < dp[j] + 1) {
          dp[i] = dp[j] + 1;
          parent[i] = j;
          if (mx < dp[i]) {
            mx = dp[i];
            mx_idx = i;
          }
        }
      }
    }
    for (int i = 0; i < mx; ++i) {
      res.push_back(nums[mx_idx]);
      mx_idx = parent[mx_idx];
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0368_Largest_Divisible_Subset.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Question_0368_Largest_Divisible_Subset {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        List<Integer> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return res;
        }
        Arrays.sort(nums);
        int[] dp = new int[nums.length], parent = new int[nums.length];
        int mx = 0, mxIndex = 0;
        for (int i = 0; i < nums.length; i++) {
            for (int j = i; j >= 0; j--) {
                if (nums[i] % nums[j] == 0 && dp[i] < dp[j] + 1) {
                    dp[i] = dp[j] + 1;
                    parent[i] = j;
                    if (dp[i] > mx) {
                        mx = dp[i];
                        mxIndex = i;
                    }
                }
            }
        }
        for (int i = 0; i < mx; i++) {
            res.add(nums[mxIndex]);
            mxIndex = parent[mxIndex];
        }
        return res;
    }
}
